fem/fespace.cpp

```diff
@@ -1632,167 +1879,406 @@ void FiniteElementSpace::Construct()
          }
       }
-      nd = V.Size() * nv + E.Size() * ne + nfd + nb;
-      dofs.SetSize(nd);
-      if (nv > 0)
+   }
+
+   if (relaxed_hp)
+   {
+      // for relaxed conformity we don't need the masters to match the minimum
+      // orders of the slaves, we can stop now
+      return;
+   }
+
+   // Iterate while minimum orders propagate by master/slave relations
+   // (and new orders also propagate from faces to incident edges).
+   // See https://github.com/mfem/mfem/pull/1423#issuecomment-638930559
+   // for an illustration of why this is necessary in hp meshes.
+   bool done;
+   do
+   {
+      done = true;
+
+      // propagate from slave edges to master edges
+      const NCMesh::NCList &edge_list = mesh->ncmesh->GetEdgeList();
+      for (const NCMesh::Master &master : edge_list.masters)
       {
-         for (k = 0; k < V.Size(); k++)
+         VarOrderBits slave_orders = 0;
+         for (int i = master.slaves_begin; i < master.slaves_end; i++)
          {
-            for (j = 0; j < nv; j++)
-            {
-               dofs[k*nv+j] = V[k]*nv+j;
-            }
+            slave_orders |= edge_orders[edge_list.slaves[i].index];
+         }
+
+         int min_order = MinOrder(slave_orders);
+         if (min_order < MinOrder(edge_orders[master.index]))
+         {
+            edge_orders[master.index] |= (VarOrderBits(1) << min_order);
+            done = false;
          }
-         nv *= V.Size();
       }
-      if (ne > 0)
+
+      // propagate from slave faces(+edges) to master faces(+edges)
+      const NCMesh::NCList &face_list = mesh->ncmesh->GetFaceList();
+      for (const NCMesh::Master &master : face_list.masters)
       {
-         // if (dim > 1)
-         for (k = 0; k < E.Size(); k++)
+         VarOrderBits slave_orders = 0;
+         for (int i = master.slaves_begin; i < master.slaves_end; i++)
          {
-            ind = fec->DofOrderForOrientation(Geometry::SEGMENT, Eo[k]);
-            for (j = 0; j < ne; j++)
+            const NCMesh::Slave &slave = face_list.slaves[i];
+            if (slave.index >= 0)
             {
-               if (ind[j] < 0)
-               {
-                  dofs[nv+k*ne+j] = -1 - ( nvdofs+E[k]*ne+(-1-ind[j]) );
-               }
-               else
+               slave_orders |= face_orders[slave.index];
+
+               mesh->GetFaceEdges(slave.index, E, ori);
+               for (int j = 0; j < E.Size(); j++)
                {
-                  dofs[nv+k*ne+j] = nvdofs+E[k]*ne+ind[j];
+                  slave_orders |= edge_orders[E[j]];
                }
             }
+            else
+            {
+               // degenerate face (i.e., edge-face constraint)
+               slave_orders |= edge_orders[-1 - slave.index];
+            }
+         }
+
+         int min_order = MinOrder(slave_orders);
+         if (min_order < MinOrder(face_orders[master.index]))
+         {
+            face_orders[master.index] |= (VarOrderBits(1) << min_order);
```

Comment:
A helper function for this kind of operation would make the code more readable.

Reply:
I disagree. These are some of the most elementary bit operations (why would they be "esoteric"?) Wrapping this in a function would be counterproductive to readability IMO.

Reply:
Is something like the following really less clear to you?
```suggestion
            AddOrder(face_orders, master.index, min_order);
```

Reply:
Yes, because it's excessive nesting. The reader has to go and check what `AddOrder` does, only to find out that it's a one-liner. Do you really go and wrap every line that looks like e.g. 
```array[index] += constant * variable?```

Reply:
Yes, I usually try to separate high level algorithms to implementation details. I don't think `AddOrder` requires to be checked, I find it self documented, how it's implemented is not really relevant. But, I'm okay to keep the code as is, it's not a big deal.